// graphql-schema-diff/dist/diff.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDiff = void 0;
const graphql_1 = require("graphql");
const utilities_1 = require("graphql/utilities");
const disparity_1 = __importDefault(require("disparity"));
const load_1 = require("@graphql-tools/load");
const url_loader_1 = require("@graphql-tools/url-loader");
const json_file_loader_1 = require("@graphql-tools/json-file-loader");
const graphql_file_loader_1 = require("@graphql-tools/graphql-file-loader");
async function getDiff(leftSchemaLocation, rightSchemaLocation, options = {}) {
    const leftSchemaOptions = {
        headers: {
            ...options.headers,
            ...(options.leftSchema && options.leftSchema.headers),
        },
        skipGraphQLImport: false,
        descriptions: false,
    };
    const rightSchemaOptions = {
        headers: {
            ...options.headers,
            ...(options.rightSchema && options.rightSchema.headers),
        },
        skipGraphQLImport: false,
        descriptions: false,
    };
    let [leftSchema, rightSchema] = await Promise.all([
        load_1.loadSchema(leftSchemaLocation, {
            loaders: [new url_loader_1.UrlLoader(), new json_file_loader_1.JsonFileLoader(), new graphql_file_loader_1.GraphQLFileLoader()],
            ...leftSchemaOptions,
        }),
        load_1.loadSchema(rightSchemaLocation, {
            loaders: [new url_loader_1.UrlLoader(), new json_file_loader_1.JsonFileLoader(), new graphql_file_loader_1.GraphQLFileLoader()],
            ...rightSchemaOptions,
        }),
    ]);
    if (!leftSchema || !rightSchema) {
        throw new Error("Schemas not defined");
    }
    if (options.sortSchema) {
        [leftSchema, rightSchema] = [
            utilities_1.lexicographicSortSchema(leftSchema),
            utilities_1.lexicographicSortSchema(rightSchema),
        ];
    }
    const [leftSchemaSDL, rightSchemaSDL] = [
        graphql_1.printSchema(leftSchema),
        graphql_1.printSchema(rightSchema),
    ];
    if (leftSchemaSDL === rightSchemaSDL) {
        return;
    }
    const diff = disparity_1.default.unified(leftSchemaSDL, rightSchemaSDL, {
        paths: [leftSchemaLocation, rightSchemaLocation],
    });
    const diffNoColor = disparity_1.default.unifiedNoColor(leftSchemaSDL, rightSchemaSDL, {
        paths: [leftSchemaLocation, rightSchemaLocation],
    });
    const dangerousChanges = graphql_1.findDangerousChanges(leftSchema, rightSchema);
    const breakingChanges = graphql_1.findBreakingChanges(leftSchema, rightSchema);
    return {
        diff,
        diffNoColor,
        dangerousChanges,
        breakingChanges,
    };
}
exports.getDiff = getDiff;
